### ISSUE TITLE
Setting image rating through manually added tag or through tag alias works, but not when tag is added by auto-tagger

### ISSUE THREAD

**Issue by Giraffaman** [shish/shimmie2#1117]
Alright here's a funny one - I wasn't sure if this really is a bug or more of a feature request since it feels like I'm abusing the rating/tag system slightly here.

 - Shimmie version:
  shimmie: "2.11.0-alpha-unknown-unknown"
  schema: 21
  php: "8.2.7"
  db: "pgsql PostgreSQL 15.6 on x86_64-pc-linux-musl, compiled by gcc (Alpine 13.2.1_git20231014) 13.2.1 20231014, 64-bit"
  os: "Linux f9a0b195826f 5.10.0-26-amd64 1 SMP Debian 5.10.197-1 (2023-09-29) x86_64"
  server: "Unit/1.32.0"

====

Trying to set up auto-tagging rules to sort of auto-rate images based on tags.

I noticed that adding a tag like "rating=e" to an image with at least one tag will set that image's rating accordingly. The "rating=e" tag disappears. This also means replacing an image's only existing tag, or uploading an image with only the "rating=e" tag will cause the "trying to set zero tags" error. 

With context in mind,  I started testing:

Scenario 1:
- define tag alias "nsfw" --> "rating=e nsfw"

Expectation:
- upload image with tag "nsfw"
- uploaded image gets tags "rating=e" and "nsfw"
- "rating=e" disappears
- tag "nsfw" remains, circumventing the "zero tags" error
- image rating is set to "Explicit"

Reality:
- works as expected

----

Scenario 2:
- define auto-tagger rule: if tag "naughty" add tag "nsfw"
- define tag alias "nsfw" --> "rating=e nsfw"

Expectation:
- upload image with tag "naughty"
- auto-tagger should see tag "naughty" and add tag "nsfw"
- tag "nsfw" should resolve to "rating=e nsfw" due to tag alias
- image should be rated "Explicit" and have tags "naughty" and "nsfw"

Reality:
- auto-tagger adds tag "nsfw"
- image ends up with tags "naughty", "nsfw", but also with tag "rating=e" still there
- image rating remains "unrated"

Clicking the image's "edit" and then "set" buttons without any changes sets image rating to "Explicit" (tag "rating=e" remains which is unexpected but acceptable).

I also tried scenario 2 with having the auto-tagger add "rating=e nsfw" directly, but same result.

It seems that tags added by auto-tagger do not trigger some kind of refresh or event being sent making Ratings aware of the added rating.

**Comment by shish** [comment]
I am also not sure whether to count that as feature or bug ^^

TBH the ability to set extra metadata bits like ratings via the tag input box is a terrible hack that a few extensions have implemented individually and inconsistently - I was on the fence about removing it completely, but this use-case does seem very useful. I wonder if there's a way to do it consistently and reliably...

**Comment by Giraffaman** [comment]
yeah wouldn't tell people to edit an image's tags to set stuff like ratings either - the dedicated menu and bulk controls for that are just easier and more intuitive to use than having to know what tagging e.g. "rating=e" does.

However, having some way to auto-rate an uploaded image, in this case based on a certain tag being present, is really neat because for the uploader, the only rule to follow is to tag "sfw"/"nsfw" (I have filters in place hiding unrated images just like nsfw for users who don't want to see that) - they don't have to upload the image and then set a rating - they just need to include that tag.

Auto-tagging would go one step further, e.g. if I see certain tags I can have the auto-tagger add appropriate tags to "auto-set" an image's rating. The mods then just need to handle the stuff that falls through the cracks. The only problem in this case right now is why tags set by the auto-tagger don't trigger the rating change while tags being replaced with their alias does.

**Comment by Giraffaman** [comment]
I [fixed](https://github.com/Giraffaman/shimmie2/commit/8d5e22b36f1513121a8292d23a73164793909004) this for me by stealing the TagSet event's constructor code from /ext/post_tags/main.php and using it in /ext/auto_tagger.php's onTagSet function to check for metatags (e.g. rating=s) and sending out a TagTermCheck event so /ext/rating can react to metatags set/added by auto_tagger.

Then I set added tag aliases like 'sfw' = 'sfw rating=s' (the rating-metatag will not get added as normal tag, therefore the extra 'sfw' which will then end up as the actual applied tag). Finally, I added an auto-tagger rule like "puppies" += "sfw".

Now, both uploads tagged with sfw (alias), or with puppies (auto-tagger) get their rating applied.
